test/json_schemer_test.rb

```diff
@@ -354,6 +354,81 @@ def test_it_returns_correct_pointers_for_ref_id
     }
   end
 
+  def test_it_returns_correct_pointers_for_a_nested_ref_id
+    ref_schema = {
+      '$id' => 'http://example.com',
+      'foo' => {
+        'bar' => {
+          '$id' => '#bar',
+          'type' => 'string'
+        }
+      }
+    }
+    root = {
+      'properties' => {
+        'a' => {
+          'properties' => {
+            'x' => { '$ref' => 'http://example.com/ref_schema.json#bar' }
+          }
+        }
+      }
+    }
+    schema = JSONSchemer.schema(
+      root,
+      ref_resolver: proc { ref_schema }
+    )
+    errors = schema.validate({ 'a' => { 'x' => 1 } }).to_a
+    assert errors.first == {
+      'data' => 1,
+      'data_pointer' => '/a/x',
+      'schema' => ref_schema['foo']['bar'],
+      'schema_pointer' => '/foo/bar',
+      'root_schema' => ref_schema,
+      'type' => 'string'
+    }
+  end
+
+  def test_it_returns_correct_pointers_for_a_nested_ref_id_with_other_pointers_to_root
+    ref_schema = {
+      '$id' => 'http://example.com',
+      'types' => {
+        'uuid' => {
+           '$id' => "#uuid",
+           'type' => 'string',
+           'pattern' => "^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$"
+        }
+      },
+      'foo' => {
+        'bar' => {
+          '$id' => '#bar',
+          'allOf' => [{ "$ref" => "#uuid"}]
+        }
+      }
+    }
+    root = {
+      'properties' => {
+        'a' => {
+          'properties' => {
+            'x' => { '$ref' => 'http://example.com/ref_schema.json#bar' }
```

Comment:
It seems like this shouldn't work.  Even though the resolver returns the ref_schema JSON, it has an `$id` for `http://example.com`, and `http://example.com#uuid`, and `http://example.com#bar`, but not `http://example.com/ref_schema.json#bar`.  (Maybe that's what you meant with the URI question?)

Reply:
The ref resolver that we pass in resolves the URI to the schema provided. All IDs are collected so specific references can be resolved. In this case, we only care about the pointer which is why we parse that out and match to the id in the referenced schema.